src/syntax_util.ml

```diff
@@ -137,34 +137,88 @@ let syntax_error_extension_node loc message =
-  while ocaml uses `not`. So, for converting between reason and ocaml syntax,
-  ocaml `not` converts to `!`, reason `!` converts to `not`. This table is
-  used later on to do the actual swapping. 
-
-  The drawback of this technique is that if you're using an identifier such as
-  `pub` in the ocaml file Foo.ml (say, let pub = 5), for a piece of Reason code
-  to interop with it, said code needs to write `Foo.method` (because of the
-  method <-> pub keyword swap). This is pretty dumb...
+(** Check to see if the string `s` is made up of `keyword` and zero or more
+    trailing `_` characters. *)
+let potentially_conflicts_with ~keyword s =
+  let s_length = String.length s in
+  let keyword_length = String.length keyword in
+  (* It can't be a match if s is shorter than keyword *)
+  s_length >= keyword_length && (
+    try
+      (* Ensure s starts with keyword... *)
+      for i = 0 to keyword_length - 1 do
+        if keyword.[i] <> s.[i] then raise Exit;
+      done;
+      (* ...and contains nothing else except trailing _ characters *)
+      for i = keyword_length to s_length - 1 do
+        if s.[i] <> '_' then raise Exit;
+      done;
+      (* If we've made it this far there's a potential conflict *)
+      true
+    with
+    | Exit -> false
+  )
+
+(** Add/remove an appropriate suffix when mangling potential keywords *)
+let string_add_suffix x = x ^ "_"
+let string_drop_suffix x = String.sub x 0 (String.length x - 1)
+
+(** What do these *_swap functions do? Here's an example: Reason code uses `!`
+    for logical not, while ocaml uses `not`. So, for converting between reason
+    and ocaml syntax, ocaml `not` converts to `!`, reason `!` converts to
+    `not`.
+
+    In more complicated cases where a keyword exists in one syntax but not the
+    other these functions translate any potentially conflicting identifier into
+    the same identifier with a suffix attached, or remove the suffix when
+    converting back.
 *)
 
-let reason_to_ml_swapping_alist = [
-  "!"       , "not";
-  "^"       , "!";
-  "++"      , "^";
-  "==="     , "==";
-  "=="      , "=";
+let reason_to_ml_swap = function
+  | "!" -> "not"
+  | "^" -> "!"
+  | "++" -> "^"
+  | "===" -> "=="
+  | "==" -> "="
   (* ===\/ and !==\/ are not representable in OCaml but
    * representable in Reason
    *)
-  "\\!=="   , "!==";
-  "\\==="   , "===";
-  "!="      , "<>";
-  "!=="     , "!=";
-  "match"   , "switch";
-  "method"  , "pub";
-  "private" , "pri";
-]
+  | "\\!==" -> "!=="
+  |  "\\===" -> "==="
+  | "!=" -> "<>"
+  | "!==" -> "!="
+  | x when (
+    potentially_conflicts_with ~keyword:"match" x
+    || potentially_conflicts_with ~keyword:"method" x
+    || potentially_conflicts_with ~keyword:"private" x) -> string_add_suffix x
+  | x when (
+    potentially_conflicts_with ~keyword:"switch_" x
+    || potentially_conflicts_with ~keyword:"pub_" x
+    || potentially_conflicts_with ~keyword:"pri_" x) -> string_drop_suffix x
```

Comment:
This is doing quite a few checks. Shouldn't it be calculated once then have the result pattern-matched on, in a nested match?

Reply:
There are a lot of checks there.  What do you mean by calculated once?